x/concentrated-liquidity/incentives.go

```diff
@@ -949,46 +961,68 @@ func (k Keeper) claimAllIncentivesForPosition(ctx sdk.Context, positionId uint64
+					err = k.communityPoolKeeper.FundCommunityPool(ctx, forfeitedIncentivesForUptime, pool.GetIncentivesAddress())
 					if err != nil {
 						return sdk.Coins{}, sdk.DecCoins{}, err
 					}
 
-					forfeitedIncentivesForPosition = forfeitedIncentivesForPosition.Add(sdk.NewDecCoinsFromCoins(collectedIncentivesForUptime...)...)
+					forfeitedIncentivesForPosition = forfeitedIncentivesForPosition.Add(sdk.NewDecCoinsFromCoins(forfeitedIncentivesForUptime...)...)
 					continue
 				}
 
 				var forfeitedIncentivesPerShare sdk.DecCoins
-				for _, coin := range collectedIncentivesForUptime {
-					// updated forfeitedIncentivesPerShare to add back = collectedIncentivesPerShare / totalSharesAccum
-					forfeitedIncentivesPerShare = append(forfeitedIncentivesPerShare, sdk.NewDecCoinFromDec(coin.Denom, coin.Amount.ToDec().Add(dust.AmountOf(coin.Denom)).Quo(totalSharesAccum)))
+				for _, forfeitedCoin := range forfeitedIncentivesForUptime {
+					// Total forfeited amount includes the dust since it gets added back into the accumulator to be distributed to other LPs.
+					forfeitedDecCoin := sdk.NewDecCoinFromDec(forfeitedCoin.Denom, forfeitedCoin.Amount.ToDec().Add(dust.AmountOf(forfeitedCoin.Denom)))
+
+					// To determine the amount of liquidity we must distribute the forfeited incentives to, we add the current tick liquidity as well as the qualifying balancer shares.
+					qualifyingLiquidity := pool.GetLiquidity().Add(qualifyingBalancerShares)
+
+					// If there is qualifying liquidity, then we can determine the amount we need to increase the accumulator by.
+					// This value gets added back into the accumulator to be distributed to other LPs.
+					if !qualifyingLiquidity.IsZero() {
+						forfeitedIncentivesPerShare = append(forfeitedIncentivesPerShare, sdk.NewDecCoinFromDec(forfeitedCoin.Denom, forfeitedDecCoin.Amount.Quo(qualifyingLiquidity)))
+					}
 
-					// convert to DecCoin to merge back with dust.
-					forfeitedIncentivesForPosition = forfeitedIncentivesForPosition.Add(sdk.NewDecCoinFromDec(coin.Denom, coin.Amount.ToDec().Add(dust.AmountOf(coin.Denom))))
+					// Track total forfeited from all uptimes for the position.
+					forfeitedIncentivesForPosition = forfeitedIncentivesForPosition.Add(forfeitedDecCoin)
 				}
 
+				// Add the incentives per share that was forfeited back into the accumulator.
 				uptimeAccum.AddToAccumulator(forfeitedIncentivesPerShare)
 				continue
 			}
 
+			// Track total incentives collected from all uptimes for the position.
 			collectedIncentivesForPosition = collectedIncentivesForPosition.Add(collectedIncentivesForUptime...)
 		}
 	}
 
+	// If the CL pool had a connected balancer pool, we had prepared the balancer pool as a full range position.
+	// We must claim and reset the balancer pool now.
+	if balancerPoolId != 0 {
+		// The uptimeAccum was modified earlier, so we must retrieve it again.
+		uptimeAccumulators, err := k.GetUptimeAccumulators(ctx, position.PoolId)
```

Comment:
Does the test behave in the same way if we don't refetch uptime accums? To my knowledge, we don't refetch when claiming balancer full range incentives so I want to make sure we're not accidentally glossing over an issue here

Reply:
+1 on this

https://github.com/osmosis-labs/osmosis/blob/c68f86668355efe5ad0313e325c92ce954f610c5/x/concentrated-liquidity/incentives.go#L446-L451

is using the modified `uptimeAccums`

Reply:
It does not behave in the same way and leads to claiming less

Reply:
I think we should actually be refetching for the call that roman linked as well